File: LeetCode/[M] - 547. Number of Provinces/solution.py
from typing import List


class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        cnt, visited = 0, set()
        for v in range(len(isConnected)):
            if v not in visited:
                cnt += 1
                self.dfs(v, visited, isConnected)
        return cnt

    def dfs(self, start_vertex, visited, isConnected):
        stack = [start_vertex]
        while stack:
            v = stack.pop()
            visited.add(v)
            for w in range(len(isConnected[v])):
                if w not in visited and isConnected[v][w]:
                    stack.append(w)

File: LeetCode/[M] - 547. Number of Provinces/solution.go
package main

func findCircleNum(isConnected [][]int) int {
	cnt, visited := 0, make(map[int]struct{})
	for i := 0; i < len(isConnected); i++ {
		if _, ok := visited[i]; !ok {
			cnt++
			dfs(i, visited, isConnected)
		}
	}
	return cnt
}

func dfs(v int, visited map[int]struct{}, isConnected [][]int) {
	stack := []int{v}
	for len(stack) != 0 {
		v = stack[0]
		stack = stack[1:]
		visited[v] = struct{}{}
		for w := 0; w < len(isConnected[v]); w++ {
			_, ok := visited[w]
			if isConnected[v][w] == 1 && !ok {
				stack = append(stack, w)
			}
		}
	}
}
